File: src/fish.py
import pygame

class Fish:
    def __init__(self, pos, level):
        self.x = pos[0]
        self.y = pos[1]
        self.y_speed = 0
        self.level = level
        self.size = 32
        self.state = 0 # 0 rest, 1 right, 2 left, 3 falling, 4 player A, 5 player B
        self.animation = 0
        self.walk_speed = 4
    
    def go_right(self):
        self.x += self.walk_speed
        self.animation = (self.animation - 1) % 40 # 40 stay on animation loop and 10 not on first sprite
        self.x_collision()

    def go_left(self):
        self.x -= self.walk_speed
        self.animation = (self.animation + 1) % 40
        self.x_collision()

    def get_tile(self, x, y):
        if (y > len(self.level) * 32):
            return True
        if self.level[y // 32][x // 32] == 'o':
            return True
        return False

    def y_collision(self):
        self.y_speed += 1
        if (self.y_speed >= 0 and (self.get_tile(self.x, self.y + 32 + self.y_speed) or self.get_tile(self.x + 31, self.y + 32 + self.y_speed))):
            tile_y = ((self.y + 32 + self.y_speed) // 32) * 32
            self.y = tile_y - 32  # Place character just above the tile
            self.y_speed = 0
            if (self.state == 3):
                self.state = 0
        self.y += self.y_speed

    def x_collision(self):
        if (((self.y_speed <= 0 and self.get_tile(self.x + 32, self.y)) or (self.y_speed >= 0 and self.get_tile(self.x + 32, self.y + 31)))):
            self.x = self.x - self.x % 32
            self.state = 2
        elif (((self.y_speed <= 0 and self.get_tile(self.x, self.y)) or (self.y_speed >= 0 and self.get_tile(self.x, self.y + 31)))):
            self.x = self.x + 32 - (self.x) % 32
            self.state = 1

    def small_col(self, pos):
        if ((((self.x + 8 <= pos[0] + 32) and (self.x + 8 >= pos[0])) or ((self.x + 23 <= pos[0] + 32) and (self.x + 24 >= pos[0])))
            and (((self.y + 9 <= pos[1] + 32) and (self.y + 9 >= pos[1])) or ((self.y + 23 <= pos[1] + 32) and (self.y + 23 >= pos[1])))):
            return True
        else:
            return False
        
    def is_hit(self, pos):
        if (self.state == 0 or self.state == 4  or self.state == 5):
            return False
        if ((((self.x + 2 <= pos[0] + 32) and (self.x + 2 >= pos[0])) or ((self.x + 29 <= pos[0] + 32) and (self.x + 30 >= pos[0])))
            and (((self.y + 3 <= pos[1] + 32) and (self.y + 3 >= pos[1])) or ((self.y + 23 <= pos[1] + 32) and (self.y + 23 >= pos[1])))):
            return True
        else:
            return False
        
    def kick(self, pos, facing):
        if (facing):
            if(not self.get_tile(pos[0] - 32, self.y)):
                self.x = pos[0] - 32
            self.state = 2
        else:
            if(not self.get_tile(pos[0] + 32, self.y)):
                self.x = pos[0] + 32
            self.state = 1

    def is_grabbed_by_a(self):
        return self.state == 4
    
    def is_grabbed_by_b(self):
        return self.state == 5
    
    def behave(self, a_pos, b_pos, a_facing, b_facing, a_grab, b_grab, a_down, a_right, a_left, b_down, b_right, b_left):
        # A method that does two things isn't ideal but for this small project it gives more clarity than splitting everything
        # I'm not using a match because some versions of python3 don't have it
        if (self.state == 0):
            if (self.small_col(a_pos) and self.small_col(b_pos)):
                return True
            if (self.small_col(a_pos)):
                if (a_grab):
                    self.state = 4
                else: 
                    self.kick(a_pos, a_facing)
            elif (self.small_col(b_pos)):
                if (b_grab):
                    self.state = 5
                else:
                    self.kick(b_pos, b_facing)
        elif (self.state == 1):
            self.go_right()
            self.y_collision()
        elif (self.state == 2):
            self.go_left()
            self.y_collision()
        elif (self.state == 3):
            self.y_collision()
        elif (self.state == 4):
            if (a_down):
                if (not self.get_tile(a_pos[0], self.y + 32)):
                    self.x = a_pos[0]
                    self.y = a_pos[1] + 32
                    self.state = 3
                else:
                    self.x = a_pos[0]
                    self.y = a_pos[1]
                    self.state = 3
            elif (a_right):
                self.kick(a_pos, False)
            elif (a_left):
                self.kick(a_pos, True)
            else:
                if (a_facing):
                    self.y = a_pos[1]
                    self.x = a_pos[0] - 20
                else:
                    self.x = a_pos[0] + 20
                    self.y = a_pos[1]
        elif (self.state == 5):
            if (b_down):
                if (not self.get_tile(a_pos[0], self.y + 32)):
                    self.x = b_pos[0]
                    self.y = b_pos[1] + 32
                    self.state = 3
                else:
                    self.x = b_pos[0]
                    self.y = b_pos[1]
                    self.state = 3
            elif (b_right):
                self.kick(b_pos, False)
            elif (b_left):
                self.kick(b_pos, True)
            else:
                if (b_facing):
                    self.y = b_pos[1]
                    self.x = b_pos[0] - 20
                else:
                    self.x = b_pos[0] + 20
                    self.y = b_pos[1]
        return False

    def draw(self, surface, sprites, a_facing, b_facing):
        # A method that does two things isn't ideal but for this small project it gives more clarity than splitting everything
        # I'm not using a match because some versions of python3 don't have it
        if (self.state == 0):
            surface.blit(sprites[0], (self.x, self.y))
        elif (self.state == 1):
            surface.blit(sprites[self.animation // 10 + 1],(self.x, self.y))
        elif (self.state == 2):
            surface.blit(sprites[self.animation // 10 + 1], (self.x, self.y))
        elif (self.state == 3):
            surface.blit(sprites[5], (self.x, self.y))
        elif (self.state == 4):
            if (a_facing):
                flipped_img = pygame.transform.flip(sprites[0], True, False)
                surface.blit(flipped_img, (self.x, self.y))
            else:
                surface.blit(sprites[0], (self.x, self.y))
        elif (self.state == 5):
            if (b_facing):
                flipped_img = pygame.transform.flip(sprites[0], True, False)
                surface.blit(flipped_img, (self.x, self.y))
            else:
                surface.blit(sprites[0], (self.x, self.y))

    def get_pos(self):
        return (self.x, self.y)
    
    def get_state(self):
        return self.state

File: src/Fish.cpp
#include "Fish.hpp"

Fish::Fish(std::pair<int,int> pos, const std::vector<std::vector<char>>& lvl)
    : x(pos.first), y(pos.second), level(lvl) {}

bool Fish::get_tile(int px, int py) const {
    if (py > (int)level.size() * TILE_SIZE) return true;
    int tx = px / TILE_SIZE, ty = py / TILE_SIZE;
    if (ty < 0 || ty >= (int)level.size())    return true;
    if (tx < 0 || tx >= (int)level[0].size()) return true;
    return level[ty][tx] == 'o';
}

void Fish::go_right() {
    x += walk_speed;
    animation = (animation - 1 + 40) % 40;
    x_collision();
}

void Fish::go_left() {
    x -= walk_speed;
    animation = (animation + 1) % 40;
    x_collision();
}

void Fish::x_collision() {
    if ((y_speed <= 0 && get_tile(x + TILE_SIZE, y)) ||
        (y_speed >= 0 && get_tile(x + TILE_SIZE, y + TILE_SIZE - 1))) {
        x     = x - (x % TILE_SIZE);
        state = FishState::Left;
    } else if ((y_speed <= 0 && get_tile(x, y)) ||
               (y_speed >= 0 && get_tile(x, y + TILE_SIZE - 1))) {
        x     = x + TILE_SIZE - (x % TILE_SIZE);
        state = FishState::Right;
    }
}

void Fish::y_collision() {
    y_speed += 1;
    if (y_speed >= 0 && (get_tile(x, y + TILE_SIZE + y_speed) ||
                          get_tile(x + TILE_SIZE - 1, y + TILE_SIZE + y_speed))) {
        int tile_y = ((y + TILE_SIZE + y_speed) / TILE_SIZE) * TILE_SIZE;
        y       = tile_y - TILE_SIZE;
        y_speed = 0;
        if (state == FishState::Falling) state = FishState::Rest;
    }
    y += y_speed;
}

bool Fish::small_col(std::pair<int,int> pos) const {
    int px = pos.first, py = pos.second;
    bool xok = ((x + 8  <= px + TILE_SIZE) && (x + 8  >= px)) ||
               ((x + 23 <= px + TILE_SIZE) && (x + 24 >= px));
    bool yok = ((y + 9  <= py + TILE_SIZE) && (y + 9  >= py)) ||
               ((y + 23 <= py + TILE_SIZE) && (y + 23 >= py));
    return xok && yok;
}

bool Fish::is_hit(std::pair<int,int> pos) const {
    if (state == FishState::Rest || state == FishState::HeldA || state == FishState::HeldB)
        return false;
    int px = pos.first, py = pos.second;
    bool xok = ((x + 2  <= px + TILE_SIZE) && (x + 2  >= px)) ||
               ((x + 29 <= px + TILE_SIZE) && (x + 30 >= px));
    bool yok = ((y + 3  <= py + TILE_SIZE) && (y + 3  >= py)) ||
               ((y + 23 <= py + TILE_SIZE) && (y + 23 >= py));
    return xok && yok;
}

void Fish::kick(std::pair<int,int> pos, bool facing) {
    if (facing) {  // player faces left - kick fish to the left
        if (!get_tile(pos.first - TILE_SIZE, y)) x = pos.first - TILE_SIZE;
        state = FishState::Left;
    } else {
        if (!get_tile(pos.first + TILE_SIZE, y)) x = pos.first + TILE_SIZE;
        state = FishState::Right;
    }
}

bool Fish::behave(std::pair<int,int> a_pos, std::pair<int,int> b_pos,
                  bool a_facing, bool b_facing,
                  bool a_grab, bool b_grab,
                  bool a_down, bool a_right, bool a_left,
                  bool b_down, bool b_right, bool b_left) {
    switch (state) {
        case FishState::Rest:
            if (small_col(a_pos) && small_col(b_pos)) return true;  // simultaneous (tie)
            if (small_col(a_pos)) {
                if (a_grab) state = FishState::HeldA; else kick(a_pos, a_facing);
            } else if (small_col(b_pos)) {
                if (b_grab) state = FishState::HeldB; else kick(b_pos, b_facing);
            }
            break;
        case FishState::Right:
            go_right();
            y_collision();
            break;
        case FishState::Left:
            go_left();
            y_collision();
            break;
        case FishState::Falling:
            y_collision();
            break;
        case FishState::HeldA:
            if (a_down) {
                x = a_pos.first; y = a_pos.second + TILE_SIZE; state = FishState::Falling;
            } else if (a_right) { kick(a_pos, false);
            } else if (a_left)  { kick(a_pos, true);
            } else {
                y = a_pos.second;
                x = a_facing ? a_pos.first - 20 : a_pos.first + 20;
            }
            break;
        case FishState::HeldB:
            if (b_down) {
                x = b_pos.first; y = b_pos.second + TILE_SIZE; state = FishState::Falling;
            } else if (b_right) { kick(b_pos, false);
            } else if (b_left)  { kick(b_pos, true);
            } else {
                y = b_pos.second;
                x = b_facing ? b_pos.first - 20 : b_pos.first + 20;
            }
            break;
    }
    return false;
}

void Fish::draw(sf::RenderTexture& target,
                const sf::Texture& tex,
                const std::vector<sf::IntRect>& frames,
                bool a_facing, bool b_facing) {
    sf::Sprite sp(tex);
    bool flip      = false;
    int  frame_idx = 0;

    switch (state) {
        case FishState::Rest:    frame_idx = 0; break;
        case FishState::Right:
        case FishState::Left:    frame_idx = animation / 10 + 1; break;
        case FishState::Falling: frame_idx = 5; break;
        case FishState::HeldA:   frame_idx = 0; flip = a_facing; break;
        case FishState::HeldB:   frame_idx = 0; flip = b_facing; break;
    }

    if (frame_idx >= (int)frames.size()) frame_idx = 0;
    sp.setTextureRect(frames[frame_idx]);

    if (flip) {
        sp.setScale(-1.f, 1.f);
        sp.setPosition((float)(x + TILE_SIZE), (float)y);
    } else {
        sp.setScale(1.f, 1.f);
        sp.setPosition((float)x, (float)y);
    }
    target.draw(sp);
}
